### ISSUE TITLE
[Chapter 10] : Exercise 10 : model restoring seems not working

### ISSUE THREAD

**Issue by mlm-root** [ageron/handson-ml2#298]
Hello guys!

I'm a bit confused about the results I get... When I run this code, the accuracy on the training and validation data is very good. When I restore the best model to evaluate it on the test data, I get 0.0971 accuracy which is very bad (see second last line) . On the other hand, if I evaluate the model on the test data without restoring the best model but using the model trained in the script, it works : accuracy = 0.9847 (see last line). It seems like the restoration is not working, or like the best model saved is finally not the best... If someone can help me ? :)

`import os
import numpy as np
import tensorflow as tf
from tensorflow import keras

(X_train_full, y_train_full), (X_test, y_test) = keras.datasets.mnist.load_data()
X_train, y_train = X_train_full[5000:] / 255., y_train_full[5000:]
X_valid, y_valid = X_train_full[:5000] / 255., y_train_full[:5000]
X_test = X_test / 255.

class ExponentialLearningRate(keras.callbacks.Callback):
	def __init__(self, factor):
		self.factor = factor
		self.rates = []
		self.losses = []
	def on_batch_end(self, batch, logs):
		self.rates.append(keras.backend.get_value(self.model.optimizer.lr))
		self.losses.append(logs["loss"])
		keras.backend.set_value(self.model.optimizer.lr, self.model.optimizer.lr * self.factor)

keras.backend.clear_session()
np.random.seed(42)
tf.random.set_seed(42)

model = keras.models.Sequential([
	keras.layers.Flatten(input_shape=[28, 28]),
	keras.layers.Dense(300, activation="relu"),
	keras.layers.Dense(100, activation="relu"),
	keras.layers.Dense(10, activation="softmax")
	])

model.compile(loss="sparse_categorical_crossentropy", optimizer=keras.optimizers.SGD(lr=2e-1), metrics=["accuracy"])
expon_lr = ExponentialLearningRate(factor=1.005)

run_index = 1 # increment this at every run
run_logdir = os.path.join("./Documents/machine_learning/chap_10", "my_mnist_logs", "run_{:03d}".format(run_index))

early_stopping_cb = keras.callbacks.EarlyStopping(patience=20)
checkpoint_cb = keras.callbacks.ModelCheckpoint("my_mnist_model.h5", save_best_only=True)
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)

history = model.fit(X_train, y_train, epochs=100, validation_data=(X_valid, y_valid), callbacks=[early_stopping_cb, checkpoint_cb, tensorboard_cb])

best_model = keras.models.load_model("my_mnist_model.h5")
best_model.evaluate(X_test, y_test) # doesn't work : accuracy = 0.0971
model.evaluate(X_test, y_test) # it works : accuracy = 0.9847
`

**Comment by ibeauregard** [comment]
This is a bug from TensorFlow; see https://github.com/tensorflow/tensorflow/issues/42045. The bug is fixed, but not in the latest version. The workaround for now is to replace 'accuracy' with 'sparse_categorical_accuracy'.

**Comment by mlm-root** [comment]
Nice, thank you!
